### DISCUSSION TITLE
the need of -fsingle-precision-constant

### DISCUSSION THREAD

**Discussion by victorandrehc** [modm-io/modm#990] [Q&A] [answered]
Hello there,
 I couldn't help but notice that you are converting every double to float, even rvalue ones,  when setting the -fsingle-precision-constant as a compilation flag. I happen to need compile time double precision in order to populate a very precise lookuptable for a temperature sensor. 

When I removed this flag from your cmake demo project I received only an error on the delay_ns.hpp which could be easily solved by casting it to float via a static_cast. I would like to know if there is any other reason behind the decision to force everything into float.

**Comment by salkinium** [comment]
This flag is added when the FPU is single precision only, so that you don't accidentally use double precision and then need a software fallback. There are also a surprising number of casts of intermediate results to double for a lot of floating point operations, which decrease performance again.
Maybe it's the wrong default, we were assuming you'd want to use the hardware FPU as much as possible and not deal with the "invisible" double casts.

**Reply by salkinium** [reply]
For double precision FPUs and no FPU, the flag is not set of course. Perhaps that's a bit weird.

**Reply by victorandrehc** [reply]
This limit the possibility to leverage double precision calculation on compile time and then cast the result do float. And leading my temperature calculations to be about 0.5ºC off. I am disabling it on my local project.

**Reply by salkinium** [reply]
Yes, and it's inconsistent behavior when you don't have a FPU at all.
The offending [line is here](https://github.com/modm-io/modm/blob/develop/src/modm/platform/core/cortex/module.lb#L372), feel free to open a PR removing it (keep the warning though).

**Reply by salkinium** [reply]
I'm getting rid of it.

**Reply by rleh** [reply]
A few floating point intense applications will break from this change when computing time was already tight before and now increases more than ten times for floating point calculations.

I would suggest to create an option in the `:platform:cortex-m` module with a nice description text so that this is more apparent to the user?

**Reply by salkinium** [reply]
Hm ok, that's a good point. You can fix it again by casting operations to floats, and the `-Wdouble-promotion` flag will point you to the problems, but it'll still be an issue for libraries you cannot change.
You can of course do `-Werror=double-promotion` if you dare ;-P

I wouldn't make it an lbuild option, since this is part of [the GCC math optimizations](https://gcc.gnu.org/wiki/FloatingPointMath), of which there are many with a lot of tradeoffs and I don't want to replicate all the docs for it. You can use the already existing collector mechanism to add compile flags already.
Instead, I would add documentation about such floating point issues into the `:build` module, since it's generic and also impacts soft floating point libraries on AVR and Cortex-M0/M3 and then point to it inside the `:platform:cortex-m` module regarding FPU precision (ie. double still using libraries).

**Reply by salkinium** [reply, marked as answer]
This flag has been removed in https://github.com/modm-io/modm/pull/1037.
